### PR TITLE
em proposal

### PR DESCRIPTION
**Pull request by sanbrock**

### CODE REVIEW

**Comment by sanbrock** [comment]
https://fairmat-experimental.github.io/nexus-fairmat-proposal/d122a69ce0c953805e60a662e9580ee2c4a6fae7/index.html

**Review by mkoennecke — COMMENTED** [review]
I do not understand the difference between NXlens and NXlens_em. NXlens seems to contain all the fields of NXlens_em too. 

Is the scan box necessary? Is this not adequately described by NeXus scan conventions?

Should NXstage not be merged with NXsample?

**Comment by woutdenolf** [comment]
Tests are failing

```
======================================================================
ERROR: test__3__contributed_definitions__NXem_nion (test_nxdl.Individual_NXDL_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/denolf/dev/nexus/definitions/utils/test_nxdl.py", line 56, in validate_xml
    result = NXDL_SCHEMA.assertValid(xml_tree)
  File "src/lxml/etree.pyx", line 3638, in lxml.etree._Validator.assertValid
lxml.etree.DocumentInvalid: Element '{http://definition.nexusformat.org/nxdl/3.1}symbols': This element is not expected. Expected is one of ( {http://definition.nexusformat.org/nxdl/3.1}doc, {http://definition.nexusformat.org/nxdl/3.1}attribute, {http://definition.nexusformat.org/nxdl/3.1}choice, {http://definition.nexusformat.org/nxdl/3.1}group, {http://definition.nexusformat.org/nxdl/3.1}field, {http://definition.nexusformat.org/nxdl/3.1}link )., line 271

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/denolf/dev/nexus/definitions/utils/test_nxdl.py", line 90, in test_wrap
    validate_xml(nxdl_file_name)
  File "/home/denolf/dev/nexus/definitions/utils/test_nxdl.py", line 61, in validate_xml
    raise NXDL_Invalid(msg)
test_nxdl.NXDL_Invalid: contributed_definitions/NXem_nion.nxdl.xml : Element '{http://definition.nexusformat.org/nxdl/3.1}symbols': This element is not expected. Expected is one of ( {http://definition.nexusformat.org/nxdl/3.1}doc, {http://definition.nexusformat.org/nxdl/3.1}attribute, {http://definition.nexusformat.org/nxdl/3.1}choice, {http://definition.nexusformat.org/nxdl/3.1}group, {http://definition.nexusformat.org/nxdl/3.1}field, {http://definition.nexusformat.org/nxdl/3.1}link )., line 271

======================================================================
ERROR: test__3__contributed_definitions__NXscanbox_em (test_nxdl.Individual_NXDL_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/denolf/dev/nexus/definitions/utils/test_nxdl.py", line 56, in validate_xml
    result = NXDL_SCHEMA.assertValid(xml_tree)
  File "src/lxml/etree.pyx", line 3638, in lxml.etree._Validator.assertValid
lxml.etree.DocumentInvalid: Element '{http://definition.nexusformat.org/nxdl/3.1}symbols': This element is not expected. Expected is one of ( {http://definition.nexusformat.org/nxdl/3.1}doc, {http://definition.nexusformat.org/nxdl/3.1}attribute, {http://definition.nexusformat.org/nxdl/3.1}choice, {http://definition.nexusformat.org/nxdl/3.1}group, {http://definition.nexusformat.org/nxdl/3.1}field, {http://definition.nexusformat.org/nxdl/3.1}link )., line 5

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/denolf/dev/nexus/definitions/utils/test_nxdl.py", line 90, in test_wrap
    validate_xml(nxdl_file_name)
  File "/home/denolf/dev/nexus/definitions/utils/test_nxdl.py", line 61, in validate_xml
    raise NXDL_Invalid(msg)
test_nxdl.NXDL_Invalid: contributed_definitions/NXscanbox_em.nxdl.xml : Element '{http://definition.nexusformat.org/nxdl/3.1}symbols': This element is not expected. Expected is one of ( {http://definition.nexusformat.org/nxdl/3.1}doc, {http://definition.nexusformat.org/nxdl/3.1}attribute, {http://definition.nexusformat.org/nxdl/3.1}choice, {http://definition.nexusformat.org/nxdl/3.1}group, {http://definition.nexusformat.org/nxdl/3.1}field, {http://definition.nexusformat.org/nxdl/3.1}link )., line 5
```

**Comment by prjemian** [comment]
The <symbols /> must appear before the <doc />.

**Comment by mkuehbach** [comment]
Wrt to mkoennecke comments from March 4th:
I do not understand the difference between NXlens and NXlens_em. NXlens seems to contain all the fields of NXlens_em too.
>> Indeed this was confusing if not incorrect. So I decided to drop NXlens_em from the proposal and propose NXlens.
There is currently only an NXxraylens but for electron microscopes using this is confusing not only because the lens designs and possible fields to store with electro-magnetic lenses, the physical characteristics are different, maybe it is possible to have at some point only one NXlens that could cater for different types of radiation. Then however the need for NXxraylens becomes questionable...

Is the scan box necessary? Is this not adequately described by NeXus scan conventions?
>> At the point when I proposed this I was not aware of NXscan. Now I inspected NXscan.
NXscan defines a generic scan-type instrument which takes a sample, scans and measures something, and the next user comes.
Indeed NXscan would apply for some modes how electron microscopes in some routine applications and 
imaging modes are used. So yes one could design the application definition for say an secondary electron image
taking electron microscope just based on NXscan and some customization but this does not cover the general
more flexible usage of electro microscopes (EM) where the EM is like an electron beam line.
Multiple detectors are used, correlatively, eventually some scan measurements on the same specimen are taken
while the user is at the instrument. So to reflect this I would had anyway to substantially reformulate the docstrings
in NXscan. I decided instead for generalizing NXem_nion into an NXem that can take account of these flexible
correlative data acquisition techniques.

Should NXstage not be merged with NXsample?
>> No I don't want this, but yes, conceptually in principle. It is a matter of debate. My aims are clear. It turned out that the discussion what exactly is the stage returned not a clear picture across electron microscopists. Usually stages nowadays are rather hierarchical arrangements of specimen fixtures. Virtually in all cases these fixture arrays are physically connected with the instrument. In that sense the stage is an entire lab of individual hierarchical components some of which user of an electron beamline find and need to have precise control over, some components the user customize, as they might bring their own fixtures which are then just mounted onto the stage.

In NeXus it seems the NXsample is the physical piece of material (the sample/specimen) + its environment, okay and this "and the environment" brings the question where do we draw the thermodynamic/conceptual boundary between where does this environment physically and conceptually end. An specimen in an EM during operation has a number of relevant environments to distinguish (near surface interaction with the chamber atmosphere, especially in environmental SEM, charging, build up of contamination) etc.
The environment in which a sample is and the stimuli which the sample is exposed, yes they could be described
 as a part of the environment but isnt then the microscope then the environment, what about electron magnetic fields.
All of these stimuli anyway need to be measured and where should these sensor readings and associated data be stored then.

**Comment by mkuehbach** [comment]
wrt to the comments of prjemian and woutdenolf:
Which tool have you used for this verification, then I can run our NXDL also through it before we upload it, thanks for point me to this both of you and the suggestions

**Comment by woutdenolf** [comment]
@mkuehbach I always do `make local` (result will be in the `./build` directory). This runs both the tests and HTML building.

File: contributed_definitions/NXaberration.nxdl.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" ?>
+<?xml-stylesheet type="text/xsl" href="nxdlformat.xsl"?>
+<definition category="base" extends="NXobject" name="NXaberration" type="group" xmlns="http://definition.nexusformat.org/nxdl/3.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.w3.org/2001/XMLSchema-instance">
+    <doc>
+         A container to store details about relevant aberrations.
```
Comment (prjemian) [inline contributed_definitions/NXaberration.nxdl.xml:5]: All these fields have terse names.  Needs some words here that these all belong to electron optics.  Documentation where these terms are defined.

File: contributed_definitions/NXebeam_column.nxdl.xml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            </doc>
+        </field>
+        <group type="NXtransformations">
+            <doc>
+                 Affine transformation which detail the arrangement in the microscope relative to
+                 the optical axis and beam path.
+            </doc>
+        </group>
+    </group>
+    <group type="NXaperture">
```
Comment (prjemian) [inline contributed_definitions/NXebeam_column.nxdl.xml:63]: might be a new base class NXaperture_em

File: contributed_definitions/NXevent_data_em.nxdl.xml [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+         microscopes &quot;&quot;&quot;perform differently&quot;&quot;&quot; across sessions, without sometimes a 
+         clear root cause. Users of the instrument may consider such conditions
+         impractical and thus either abort their session or try to bring the 
+         microscope first into a state where conditions are considered stable and
+         of high enough qualiy to collect data.
+         
+         In all these cases it is practical to store time-dependent data of the
+         instrument state not in the respective instrument component groups
+         of the top-level NXinstrument but in a sort of event data log.
+         This is the idea behind this NX&quot;#&quot;em&quot;#&quot;_snapshot.
```
Comment (prjemian) [inline contributed_definitions/NXevent_data_em.nxdl.xml:59]: formatting is off

File: contributed_definitions/NXevent_data_em.nxdl.xml [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+         details about its components, locations and eventually applied stimuli and 
+         positioning of the specimen. Some snapshot or integrated data from this stream
+         are relevant to understand the signal genesis. In such a generic case
+         it might be necessary to sync these streaming data with those intervals
+         in time when specific measurements are taken (spectra collected, 
+         images taken, diffraction images indexed on-the-fly).
+         
+         Theoretically, an instrument and specimen should be considered as dynamic.
+         Scientists often report or feel (difficult to quantify) observations that
+         microscopes &quot;&quot;&quot;perform differently&quot;&quot;&quot; across sessions, without sometimes a
```
Comment (prjemian) [inline contributed_definitions/NXevent_data_em.nxdl.xml:50]: quotation marks need attention here

File: contributed_definitions/NXibeam_column.nxdl.xml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                 the optical axis and beam path.
+            </doc>
+        </group>
+    </group>
+    <group type="NXaperture"/>
+    <group type="NXlens_em"/>
+    <group type="NXsensor"/>
+    <group type="NXbeam">
+        <doc>
+             Individual ocharacterization results for the position, shape,
```
Comment (prjemian) [inline contributed_definitions/NXibeam_column.nxdl.xml:91]: spelling

File: contributed_definitions/NXibeam_column.nxdl.xml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+         
+         The fact that an electron microscope with FIB capabilities has a 
+         second gun with own relevant control circuits, focusing lenses, and
+         other components warrants an own base class to group these components
+         and distinguish them from the lenses and components of the
+         electron column.
+         
+         For more details about the relevant physics and application examples
+         consult the literature, for example:
+         * [L. A. Giannuzzi et al.](https://doi.org/10.1007/b101190)
```
Comment (prjemian) [inline contributed_definitions/NXibeam_column.nxdl.xml:21]: blank lines before and after bullet list

File: contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+        </symbol>
+    </symbols>
+    <doc>
+         Electron backscatter diffraction (EBSD) Kikuchi pattern
+         
+         The container can also store data related to a post-processing of these
+         Kikuchi pattern, which is the backbone of orientation microscopy
+         especially in materials science and materials engineering.
+         
+         Based on a fuse of the [h5ebsd](https://doi.org/10.1186/2193-9772-3-4)
```
Comment (prjemian) [inline contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml:28]: markdown-style URL convert to reST (as before)

File: contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+         of the DREAM.3D community and the open H5OINA format of Oxford Instruments
+         [H5OINA](https://doi.org/10.1017/S1431927621006103).
+         
+         EBSD can be used, usually FIB/SEM microscopes, provided for three-dimensional
+         orientation microscopy. So-called serial section analyses. For a detailed
+         overview of these techniques see e.g.:
+         * [M. A. Groeber](https://doi.org/10.1186/2193-9772-3-5)
```
Comment (prjemian) [inline contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml:35]: bullet list needs blank line before and after

File: contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+                     Minimum number of bands required to index the pattern
+                </doc>
+                <dimensions rank="1">
+                    <dim index="1" value="n_p"/>
+                </dimensions>
+            </field>
+            <field name="status" type="NX_NUMBER" units="NX_UNITLESS">
+                <doc>
+                     Which return value did the indexing algorithm yield for each pattern.
+                     * Details about bad pixels
```
Comment (prjemian) [inline contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml:253]: bullet list needs blank line before and after

File: contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+                </field>
+                <field name="phase_name" type="NX_CHAR">
+                    <doc>
+                         Name of the phase/alias.
+                    </doc>
+                </field>
+                <field name="number_of_reflectors" type="NX_UINT"/>
+                <field name="miller_indices" type="NX_NUMBER" units="NX_UNITLESS">
+                    <doc>
+                         Miller indices &quot;(&quot;hkl&quot;)&quot;&quot;[&quot;uvw&quot;]&quot;.
```
Comment (prjemian) [inline contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml:226]: needs attention to formatting here

File: contributed_definitions/NXinteraction_vol_em.nxdl.xml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+         represented ideally as trajectory data from the computer simulation.
+         * An explicit description is via an iso-contour surface using either
+         a simulation grid or a triangulated surface mesh of the approximated
+         iso-contour surface evaluated at specific threshold values.
+         Iso-contours could be computed from electron or particle fluxes through
+         an imaginary control surface (the iso-surface). 
+         Threshold values can be defined by particles passing through a unit control
+         volume (electrons) or energy-levels (e.g. the case of X-rays).
+         Details depend on the model.
+         * Another explicit description is via theoretical models which may
```
Comment (prjemian) [inline contributed_definitions/NXinteraction_vol_em.nxdl.xml:25]: bullet list needs blank line before and after

File: contributed_definitions/NXscanbox_em.nxdl.xml [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+         Scan box and coils which deflect the electron beam in a controlled manner.
+         
+         In electron microscopy, the scan box is instructed by the microscope
+         control software. This component directs the probe to controlled locations
+         according to a scan scheme and plan.
+    </doc>
+    <field name="calibration_style" type="NX_CHAR"/>
+    <field name="center" type="NX_NUMBER" units="NX_ANY"/>
+    <field name="flyback_time" type="NX_FLOAT" units="NX_TIME"/>
+    <field name="field_of_view" type="NX_FLOAT" units="NX_LENGTH"/>
```
Comment (prjemian) [inline contributed_definitions/NXscanbox_em.nxdl.xml:14]: duplicate from NXoptical_system

File: contributed_definitions/NXem.nxdl.xml [hunk 13]
```diff
@@ -0,0 +1,723 @@
...
+         The development in the past has shown that these activities have led to a
+         zoo of schemata and implementations of these into data and file formats.
+         There is nothing which prevents the communities to make these schemata
+         open and interoperable. Open here means specifically not that all data are
+         compliant with/or use the schema and have to end up in the open-source domain.
+         There can be embargo periods first of all. Open means that the metadata and 
+         associated schemata are documented in a manner that as many details as 
+         possible are open in the sense that others can understand what the 
+         (meta)data mean conceptually. 
+         The [FAIR](https://doi.org/10.1038/sdata.2016.18) principles guide all
```
Comment (prjemian) [inline contributed_definitions/NXem.nxdl.xml:73]: markdown-style reference needs to be converted to reST-style

File: contributed_definitions/NXem.nxdl.xml [hunk 14]
```diff
@@ -0,0 +1,723 @@
...
+         a time stamp and link to the specific settings of the microscope at that
+         point in time when the image was taken. The reason is that EMs can be highly
+         dynamic and thus physically illuminate the specimen differently or show 
+         drift during signal acquisition, to name but a few effects.
+         What constitutes a single EM experiment/measurement? This can be the
+         collecting of a single diffraction pattern with a scanning TEM (STEM), 
+         taking of a secondary electron image for fracture analysis, taking a set of
+         EBSD line scan and surface mappings in an SEM, or ion-beam-milling of a
+         specimen in preparation for an atom probe experiment.
+         * NXmonitor;
```
Comment (prjemian) [inline contributed_definitions/NXem.nxdl.xml:213]: bullet list needs blank lines before & after

File: contributed_definitions/NXem.nxdl.xml [hunk 15]
```diff
@@ -0,0 +1,723 @@
...
+         the data are stored. NeXus specifies only the data model, i.e. the terms
+         and their relations. These descriptions can be implemented and stored in
+         Json, HDF5, XML, or HSDS, file storage, or even other formats, although
+         HDF5 is the most commonly used.
+         * Consumable results of EM characterization tasks are usually a sub-set of
+         data artifacts, as there is not an infinite amount of possible beam-specimen
+         interactions.
+         * Images of electron counts detected in specific operation modes
+         (bright field, dark field in TEM, secondary/back-scattered, Kikuchi in SEM)
+         * Spectra (X-ray quanta or auger electron counts)
```
Comment (prjemian) [inline contributed_definitions/NXem.nxdl.xml:257]: bullet list needs blank lines before & after

File: contributed_definitions/NXem.nxdl.xml [hunk 16]
```diff
@@ -0,0 +1,723 @@
...
+         taken: secondary, back-scattered, electron image, diffraction pattern,
+         EELS spectra, composition or orientation mappings.
+         
+         There are a few design choices to consider with sub-ordinate groups:
+         * Above images, spectra, and mappings should be stored as NXdata instances,
+         ideally formatted in such a way that they can be displayed with 
+         visualization software that can be specific for the file format in which 
+         the data are stored. NeXus specifies only the data model, i.e. the terms
+         and their relations. These descriptions can be implemented and stored in
+         Json, HDF5, XML, or HSDS, file storage, or even other formats, although
```
Comment (prjemian) [inline contributed_definitions/NXem.nxdl.xml:250]: change `Json` to `JSON`

**Review by prjemian — APPROVED** [review]
LGTM as contributed

File: contributed_definitions/NXaberration.nxdl.xml (hunk 1 above)
Comment (mkuehbach) [reply contributed_definitions/NXaberration.nxdl.xml:5]: added literature and docstrings for details

File: contributed_definitions/NXebeam_column.nxdl.xml (hunk 2 above)
Comment (mkuehbach) [reply contributed_definitions/NXebeam_column.nxdl.xml:63]: fixed

File: contributed_definitions/NXevent_data_em.nxdl.xml (hunk 4 above)
Comment (mkuehbach) [reply contributed_definitions/NXevent_data_em.nxdl.xml:50]: formatting fixed, will carry this as mentioned also through for the other places where bullet point lists when havoc, thank you prjemian

File: contributed_definitions/NXibeam_column.nxdl.xml (hunk 5 above)
Comment (mkuehbach) [reply contributed_definitions/NXibeam_column.nxdl.xml:91]: fixed

File: contributed_definitions/NXibeam_column.nxdl.xml (hunk 6 above)
Comment (mkuehbach) [reply contributed_definitions/NXibeam_column.nxdl.xml:21]: changed, also for all other enumerations and itemize lists

File: contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml (hunk 7 above)
Comment (mkuehbach) [reply contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml:28]: fixed

File: contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml (hunk 9 above)
Comment (mkuehbach) [reply contributed_definitions/NXimage_set_em_kikuchi.nxdl.xml:253]: fixed

File: contributed_definitions/NXem.nxdl.xml (hunk 16 above)
Comment (mkuehbach) [reply contributed_definitions/NXem.nxdl.xml:250]: fixed

File: contributed_definitions/NXem.nxdl.xml (hunk 14 above)
Comment (mkuehbach) [reply contributed_definitions/NXem.nxdl.xml:213]: fixed

File: contributed_definitions/NXem.nxdl.xml (hunk 13 above)
Comment (mkuehbach) [reply contributed_definitions/NXem.nxdl.xml:73]: fixed

File: contributed_definitions/NXevent_data_em.nxdl.xml (hunk 3 above)
Comment (mkuehbach) [reply contributed_definitions/NXevent_data_em.nxdl.xml]: fixed

**Review by mkuehbach — APPROVED** [review]

**Review by mkuehbach — APPROVED** [review]

File: contributed_definitions/NXaberration.nxdl.xml (hunk 1 above)
Comment (prjemian) [reply contributed_definitions/NXaberration.nxdl.xml]: I'm not able to see any additions from GitHub's viewer.  However, there are changes visible in my local clone of this repo, once I pull the changes.

File: contributed_definitions/NXcoordinate_system_set.nxdl.xml [hunk 17]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" ?>
+<?xml-stylesheet type="text/xsl" href="nxdlformat.xsl"?>
+<definition category="base" extends="NXobject" name="NXcoordinate_system_set" type="group" xmlns="http://definition.nexusformat.org/nxdl/3.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.w3.org/2001/XMLSchema-instance">
```
Comment (prjemian) [inline contributed_definitions/NXcoordinate_system_set.nxdl.xml:3]: change

```xml
    xsi:schemaLocation="http://www.w3.org/2001/XMLSchema-instance"
```

to

```xml
    xsi:schemaLocation="http://definition.nexusformat.org/nxdl/3.1 ../nxdl.xsd"
```

Comment (prjemian) [reply contributed_definitions/NXcoordinate_system_set.nxdl.xml:3]: Check all nxdl.xml files in this PR (#996), all #995, #997, and #998.  Revise each as needed.

Comment (sanbrock) [reply contributed_definitions/NXcoordinate_system_set.nxdl.xml:3]: All other issues are fixed in this and the other PRs, this issue will also be fixed tomorrow.

Comment (sanbrock) [reply contributed_definitions/NXcoordinate_system_set.nxdl.xml]: this and all the other related PRs are now updated and ready for being merged. #995 #996 #997 #998.

**Pull request merged**